#!/bin/bash
#SBATCH --job-name=g7445_diploids_consensus
#SBATCH --partition=defq
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=1
#SBATCH --mem=16g
#SBATCH --time=8:00:00
#SBATCH --output=/gpfs01/home/sbzsmb/OandE/%x.out
#SBATCH --error=/gpfs01/home/sbzsmb/OandE/%x.err
source $HOME/.bash_profile
conda activate ngs_pipe_env_gatk4
ml bcftools-uoneasy/1.17-GCC-12.2.0
gatk SelectVariants -R /gpfs01/home/mbzly/2022.Cochlearia.MS1/ngs_pipe/ionops/ref.gen/C_excelsa_V5.fasta -V /gpfs01/home/mbzly/2022.Cochlearia.MS1/ngs_pipe/08.recombined/recombined.vcf.gz --restrict-alleles-to BIALLELIC --max-nocall-fraction 0.2 -select "AF > 0.5" -sn bnk_21 -sn cha_1 -sn cha_2 -sn jor_1 -sn jor_12 -sn jor_13 -sn jor_3 -sn LAB_004 -sn lab_A -sn lab_B -sn LAB_300 -sn lab_D -sn LAB_400 -sn lab_E -sn LAB_500 -sn NEN_001 -sn NEN_003 -sn NEN_200 -sn NEN_300 -sn nen_4 -sn nen_5 -sn nen_6 -sn odn_10 -sn odn_2 -sn odn_4 -sn odn_5 -sn odn_6 -sn odn_7 -sn odn_9 -L Cexcelsa_scaf_4:35100912-35101038 -L Cexcelsa_scaf_4:35101122-35101211 -L Cexcelsa_scaf_4:35101400-35101518 -L Cexcelsa_scaf_4:35101660-35101773 -L Cexcelsa_scaf_4:35101851-35101932 -L Cexcelsa_scaf_4:35102121-35102188 -L Cexcelsa_scaf_4:35102266-35102340 -L Cexcelsa_scaf_4:35102425-35102570 -L Cexcelsa_scaf_4:35102651-35102753 -L Cexcelsa_scaf_4:35102848-35102928 -L Cexcelsa_scaf_4:35103056-35103124 -L Cexcelsa_scaf_4:35103206-35103265 -L Cexcelsa_scaf_4:35103340-35103507 -L Cexcelsa_scaf_4:35103838-35103894 -L Cexcelsa_scaf_4:35103978-35104220 -L Cexcelsa_scaf_4:35104295-35104360 -L Cexcelsa_scaf_4:35104433-35104615 -L Cexcelsa_scaf_4:35104719-35104823 -L Cexcelsa_scaf_4:35105015-35105098 -L Cexcelsa_scaf_4:35105213-35105359 -L Cexcelsa_scaf_4:35105501-35105614 -L Cexcelsa_scaf_4:35105700-35105786 -L Cexcelsa_scaf_4:35105886-35106089 -L Cexcelsa_scaf_4:35106192-35106338 -L Cexcelsa_scaf_4:35106692-35106898 -L Cexcelsa_scaf_4:35106997-35107086 -L Cexcelsa_scaf_4:35107150-35107290 -L Cexcelsa_scaf_4:35107540-35107637 -L Cexcelsa_scaf_4:35107732-35107810 -L Cexcelsa_scaf_4:35107916-35108080 -L Cexcelsa_scaf_4:35108184-35108414 -O /gpfs01/home/sbzsmb/Cochlearia/AlphaFold/Consensus/g7445_diploids_biallelic.vcf
gatk SelectVariants -R /gpfs01/home/mbzly/2022.Cochlearia.MS1/ngs_pipe/ionops/ref.gen/C_excelsa_V5.fasta -V /gpfs01/home/mbzly/2022.Cochlearia.MS1/ngs_pipe/08.recombined/recombined.vcf.gz --restrict-alleles-to MULTIALLELIC --max-nocall-fraction 0.2 --exclude-non-variants -sn bnk_21 -sn cha_1 -sn cha_2 -sn jor_1 -sn jor_12 -sn jor_13 -sn jor_3 -sn LAB_004 -sn lab_A -sn lab_B -sn LAB_300 -sn lab_D -sn LAB_400 -sn lab_E -sn LAB_500 -sn NEN_001 -sn NEN_003 -sn NEN_200 -sn NEN_300 -sn nen_4 -sn nen_5 -sn nen_6 -sn odn_10 -sn odn_2 -sn odn_4 -sn odn_5 -sn odn_6 -sn odn_7 -sn odn_9 -L Cexcelsa_scaf_4:35100912-35101038 -L Cexcelsa_scaf_4:35101122-35101211 -L Cexcelsa_scaf_4:35101400-35101518 -L Cexcelsa_scaf_4:35101660-35101773 -L Cexcelsa_scaf_4:35101851-35101932 -L Cexcelsa_scaf_4:35102121-35102188 -L Cexcelsa_scaf_4:35102266-35102340 -L Cexcelsa_scaf_4:35102425-35102570 -L Cexcelsa_scaf_4:35102651-35102753 -L Cexcelsa_scaf_4:35102848-35102928 -L Cexcelsa_scaf_4:35103056-35103124 -L Cexcelsa_scaf_4:35103206-35103265 -L Cexcelsa_scaf_4:35103340-35103507 -L Cexcelsa_scaf_4:35103838-35103894 -L Cexcelsa_scaf_4:35103978-35104220 -L Cexcelsa_scaf_4:35104295-35104360 -L Cexcelsa_scaf_4:35104433-35104615 -L Cexcelsa_scaf_4:35104719-35104823 -L Cexcelsa_scaf_4:35105015-35105098 -L Cexcelsa_scaf_4:35105213-35105359 -L Cexcelsa_scaf_4:35105501-35105614 -L Cexcelsa_scaf_4:35105700-35105786 -L Cexcelsa_scaf_4:35105886-35106089 -L Cexcelsa_scaf_4:35106192-35106338 -L Cexcelsa_scaf_4:35106692-35106898 -L Cexcelsa_scaf_4:35106997-35107086 -L Cexcelsa_scaf_4:35107150-35107290 -L Cexcelsa_scaf_4:35107540-35107637 -L Cexcelsa_scaf_4:35107732-35107810 -L Cexcelsa_scaf_4:35107916-35108080 -L Cexcelsa_scaf_4:35108184-35108414 -O /gpfs01/home/sbzsmb/Cochlearia/AlphaFold/Consensus/g7445_diploids_multiallelic.vcf
bgzip /gpfs01/home/sbzsmb/Cochlearia/AlphaFold/Consensus/g7445_diploids_biallelic.vcf
tabix -p vcf /gpfs01/home/sbzsmb/Cochlearia/AlphaFold/Consensus/g7445_diploids_biallelic.vcf.gz
samtools faidx /gpfs01/home/mbzly/2022.Cochlearia.MS1/ngs_pipe/ionops/ref.gen/C_excelsa_V5.fasta Cexcelsa_scaf_4:35100912-35101038 Cexcelsa_scaf_4:35101122-35101211 Cexcelsa_scaf_4:35101400-35101518 Cexcelsa_scaf_4:35101660-35101773 Cexcelsa_scaf_4:35101851-35101932 Cexcelsa_scaf_4:35102121-35102188 Cexcelsa_scaf_4:35102266-35102340 Cexcelsa_scaf_4:35102425-35102570 Cexcelsa_scaf_4:35102651-35102753 Cexcelsa_scaf_4:35102848-35102928 Cexcelsa_scaf_4:35103056-35103124 Cexcelsa_scaf_4:35103206-35103265 Cexcelsa_scaf_4:35103340-35103507 Cexcelsa_scaf_4:35103838-35103894 Cexcelsa_scaf_4:35103978-35104220 Cexcelsa_scaf_4:35104295-35104360 Cexcelsa_scaf_4:35104433-35104615 Cexcelsa_scaf_4:35104719-35104823 Cexcelsa_scaf_4:35105015-35105098 Cexcelsa_scaf_4:35105213-35105359 Cexcelsa_scaf_4:35105501-35105614 Cexcelsa_scaf_4:35105700-35105786 Cexcelsa_scaf_4:35105886-35106089 Cexcelsa_scaf_4:35106192-35106338 Cexcelsa_scaf_4:35106692-35106898 Cexcelsa_scaf_4:35106997-35107086 Cexcelsa_scaf_4:35107150-35107290 Cexcelsa_scaf_4:35107540-35107637 Cexcelsa_scaf_4:35107732-35107810 Cexcelsa_scaf_4:35107916-35108080 Cexcelsa_scaf_4:35108184-35108414 | bcftools consensus -s - /gpfs01/home/sbzsmb/Cochlearia/AlphaFold/Consensus/g7445_diploids_biallelic.vcf.gz -o /gpfs01/home/sbzsmb/Cochlearia/AlphaFold/Consensus/g7445_diploids_biallelic.fasta
